hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,346 @@
+
+    // inputs of conditional propagate into case nodes
+    out_wire_value(c, i_node, OutgoingPort::from(out_p), v) <--
+      case_node(c, cond, case_index, case),
+      io_node(c, case, i_node, IO::Input),
+      node_in_value_row(c, cond, in_row),
+      let conditional = c.get_optype(*cond).as_conditional().unwrap(),
+      if let Some(fields) = in_row.unpack_first(*case_index, conditional.sum_rows[*case_index].len()),
+      for (out_p, v) in fields.enumerate();
+
+    // outputs of case nodes propagate to outputs of conditional *if* case reachable
+    out_wire_value(c, cond, OutgoingPort::from(o_p.index()), v) <--
+      case_node(c, cond, _, case),
+      case_reachable(c, cond, case),
+      io_node(c, case, o, IO::Output),
+      in_wire_value(c, o, o_p, v);
+
+    relation case_reachable(C, Node, Node);
+    case_reachable(c, cond, case) <-- case_node(c,cond,i,case),
+        in_wire_value(c, cond, IncomingPort::from(0), v),
+        if v.supports_tag(*i);
+
+    // CFG
+    relation cfg_node(C, Node);
+    relation dfb_block(C, Node, Node);
+    cfg_node(c,n) <-- node(c, n), if c.get_optype(*n).is_cfg();
+    dfb_block(c,cfg,blk) <-- cfg_node(c, cfg), for blk in c.children(*cfg), if c.get_optype(blk).is_dataflow_block();
+
+    // Reachability
+    relation bb_reachable(C, Node, Node);
+    bb_reachable(c, cfg, entry) <-- cfg_node(c, cfg), if let Some(entry) = c.children(*cfg).next();
+    bb_reachable(c, cfg, bb) <-- cfg_node(c, cfg),
+        bb_reachable(c, cfg, pred),
+        io_node(c, pred, pred_out, IO::Output),
+        in_wire_value(c, pred_out, IncomingPort::from(0), predicate),
+        for (tag, bb) in c.output_neighbours(*pred).enumerate(),
+        if predicate.supports_tag(tag);
+
+    // Where do the values "fed" along a control-flow edge come out?
+    relation _cfg_succ_dest(C, Node, Node, Node);
+    _cfg_succ_dest(c, cfg, blk, inp) <-- dfb_block(c, cfg, blk), io_node(c, blk, inp, IO::Input);
+    _cfg_succ_dest(c, cfg, exit, cfg) <-- cfg_node(c, cfg), if let Some(exit) = c.children(*cfg).nth(1);
+
+    // Inputs of CFG propagate to entry block
+    out_wire_value(c, i_node, OutgoingPort::from(p.index()), v) <--
+        cfg_node(c, cfg),
+        if let Some(entry) = c.children(*cfg).next(),
+        io_node(c, entry, i_node, IO::Input),
+        in_wire_value(c, cfg, p, v);
+
+    // Outputs of each reachable block propagated to successor block or (if exit block) then CFG itself
+    out_wire_value(c, dest, OutgoingPort::from(out_p), v) <--
+        dfb_block(c, cfg, pred),
+        bb_reachable(c, cfg, pred),
+        let df_block = c.get_optype(*pred).as_dataflow_block().unwrap(),
+        for (succ_n, succ) in c.output_neighbours(*pred).enumerate(),
+        io_node(c, pred, out_n, IO::Output),
+        _cfg_succ_dest(c, cfg, succ, dest),
+        node_in_value_row(c, out_n, out_in_row),
+        if let Some(fields) = out_in_row.unpack_first(succ_n, df_block.sum_rows.get(succ_n).unwrap().len()),
```

Comment:
This grabs the row in input values for the node and only looks at the first. It could get the first value directly by using `in_wire_value`. There's some other places where this is done as well.

Reply:
I see! I think it's clear enough now.